### ISSUE TITLE
Listing and selecting buffers and files (e.g. NERDTree)

### ISSUE THREAD

**Issue by akjanik** [mawww/kakoune#1507]
Hi,

I recently discovered Kakoune and decided to try this "better vim". At first glance looks nice. However, before I start using it for daily basis, I need to ensure that it have some basic stuff I'm so used to that I can't abandon.

The first one is something like NERDTree plugin. I didn't find equivalent so far, is there any way to work with project, in which I need to quickly open/close files in different locations?

Much more important is listing buffers. In VIM I can run `:ls` to get list of files with associated buffer numbers and then `:b<number>` to open file `<number>`. I found commands `:bn` and `:bp` for next/previous buffer, but honestly, having opened a lot of files, I can't iterate over them to get the desired one (and I don't even when I reach it).

Cheers.

**Comment by Delapouite** [comment]
Hello

When discovering a new code base and need to explore it for a while before getting used to it, I often rely on ranger: https://github.com/ranger/ranger
Then, I rely on fzf to open the file I need: https://github.com/junegunn/fzf

Both of these tools are external to kakoune, so for a good integration, you can use the guide found on this page: https://github.com/mawww/kakoune/wiki/Open-files-in-specific-client

Finally when I have a lot of buffers open, I use the builtin `:buffer` command which has a nice autocomplete, or this [plugin](https://github.com/Delapouite/kakoune-buffers) for extra commands.

**Comment by akjanik** [comment]
Tha last paragraph seems to be answer to my issue. I am very surprise this is not "core" functionality of editor, as I can hardly imagine serious workflow with many files without this feature.

However, I created directory `.config/kak/autoload` and copied your script here, but it doesn't work completely - command `list-buffers` do nothing, even error is not returned. Same if I choose `b` after `:modes-buffers` (which work in most cases). `debug` is being filled with (Ubuntu 16.04, terminator):
`shell stderr: <<<
/bin/sh: 3: Bad substitution
>>>
`
or 
`
/bin/sh: 1 : [: := unexpected operator
`
And I had to type it by hand, because for some reason copying to clipboard from console (ctrl+shift+c) does not work...

I think I'll stay with (neo)vim. Project looks nice, but it need to mature a lot before I start using it.

Thx for your help anyway!

**Comment by mawww** [comment]
I am not sure whats missing to just typing `:b`, you'll get all the buffers displayed in the auto completion menu, and can select the one you want just by using fuzzy matching to reduce that completion list.

We dont have a built-in buffer listing command because its not needed, the buffer list gets automatically displayed in contexts where you need it, such as when switching buffers. `:b<space>` will show it.

**Comment by mawww** [comment]
I dont think there is much we need to change here, we have the feature, in a more interactive way than (neo)vim.

**Comment by 42xel** [comment]
While `:b<space>` covers all daily use cases, I don't think it's strictly better than a buffer list in a scratch buffer. 

A buffer list can be interacted programmatically, including with the full power of Kakoune. 

Here is a usecase I just ran into: using a macro, I chained created a dozen and a half buffers corresponding to new files, with dynamic names. I then realized most of these filenames have slashes, interpreted as directories, which don't exist. Fortunately, it makes sense to create the corresponding directory arborescence. In hindsight, my macro should have handled dir creation, but as it stands, I have twenty buffers I cannot save. With a buffer list, a solution to my problem would be as simple as selecting each line and applying `dirname` then `mkdir -p` .
